File: scripts/recalculatePath.py
from error_calc import errorCalc
from math import *
from Point import Point
from vehicleState import VehicleState
import rospy
from custom_msgs.msg import Path, Position
from helper_functions import *

class recalculatePath:

    def __init__(self, speed, trackChecker):
        self.speed = speed
        self.trackChecker = trackChecker

        self.path_pub = rospy.Publisher('possible_path', Path, queue_size=10)


    def calculate_path(self, startPoint, snd_to_last_end, endPoint, dt, theta1, theta2, totError, ec):

        self.start_theta1 = theta1
        self.start_theta2 = theta2
        self.errorList = {}

        self.theta1 = theta1 #start angle for header
        self.theta2 = theta2 #start angle for trailer
        self.dt = dt #the delta time used for kinematic model, basicly the path step size
        self.front_ec_i = 1
        self.back_ec_i = 1
        self.ec = ec
        self.path = ([],[], self.ec) #first list is the fromPoints, second list is the solution path

        self.lowest_error = totError
        self.totError = 0
        dd = self.speed * self.dt

        self.pos = startPoint
        self.fromPoints = {}
        self.toVisit = []
        self.visited = set([]) #the points we have visited
        #add all possible pathes for the first point before looping
        self.addPossiblePathes()

        print len(self.toVisit)
        count = 0
        while len(self.toVisit)>=0:
            count= count+1
            #loop until all possible nodes have been visited
            while True:
                if len(self.toVisit)==0:
                    print "reached End"
                    print "lowest error: ", self.lowest_error
                    print "start error: ", totError
                    print "starPoint: ", startPoint.x, startPoint.y
                    print "endPoint: ", endPoint.x, endPoint.y
                    return self.path
                ((x,y),t1, t2, err, toterr, new_front_ec_i, new_back_ec_i) = self.toVisit.pop()
                #round to make it faster, not having to visit as many nodes that are similar

                ((round_x, round_y), round_theta1, round_theta2) = rounding(x, y, t1, t2)
                prev_err = self.errorList[((round_x, round_y), round_theta1, round_theta2)]
                if ((round_x,round_y),round_theta1, round_theta2) not in self.visited: #or prev_err>toterr:
                    break
            #found new node to visit
            self.pos = Point(x,y) # get the toPoint
            #print x,y,err, toterr
            self.front_ec_i = new_front_ec_i
            self.back_ec_i = new_back_ec_i
            self.theta1= t1
            self.theta2= t2
            self.totError = toterr
            #check if we have reached the end
            dist = sqrt( (endPoint.x - x)**2 + (endPoint.y - y)**2 )
            if self.ec.isAboveEnd(snd_to_last_end, endPoint, self.pos) and dist <1*self.dt and self.ec.isAtEnd(self.front_ec_i): #checks if we are above a line of the two last points
                #reached a sloution, gather the path if we are on the optimal path
                count = count+1
                print count
                #rospy.sleep(0.2)
                (new_point, nt1, nt2) = calculate_steering(radians(MAX_RIGHT_ANGLE), radians(MAX_LEFT_ANGLE), dd, 10, 0, self.pos, self.theta1, self.theta2, self.ec, self.front_ec_i)
                (nerror, n_i)= self.ec.calculateError(new_point, self.front_ec_i)
                self.path_pub.publish(Path(self.gather_x_y_path(startPoint, new_point, nt1, nt2, nerror)))
                if abs(nerror)< 1 and toterr < self.lowest_error:
                    print "found better solution"
                    self.path = self.gatherPath(startPoint, new_point, nt1, nt2, nerror, err)
                    self.lowest_error = toterr
                #mark visited
                self.visited.add(((self.pos.x, self.pos.y),self.theta1, self.theta2))

            else:
                #we have not yet found a solution, search for new possible nodes
                self.addPossiblePathes()

                ((round_x, round_y), round_th1, round_th2) = rounding(self.pos.x, self.pos.y, self.theta1, self.theta2)
                self.visited.add(((round_x, round_y),round_theta1, round_theta2))

    def addPossiblePathes(self):
        dd = self.speed * self.dt

        steering_angle_rad = radians(0) #strait

        (to_point_strait, strait_theta1, strait_theta2) = calculateNextState(self.theta1, self.theta2, self.pos, dd, steering_angle_rad)

        #going right
        steering_angle_rad = radians(MAX_LEFT_ANGLE) #max right angle

        (to_point_right,right_theta1,right_theta2) = calculateNextState(self.theta1, self.theta2, self.pos, dd, steering_angle_rad)

        #going left
        steering_angle_rad = radians(MAX_RIGHT_ANGLE) #max left angle

        (to_point_left,left_theta1, left_theta2) = calculateNextState(self.theta1, self.theta2, self.pos, dd, steering_angle_rad)

        #finding optimal path
        (to_point_optimal, optimal_theta1, optimal_theta2) = calculate_steering(radians(MAX_RIGHT_ANGLE), radians(MAX_LEFT_ANGLE), dd, 10, 0, self.pos, self.theta1, self.theta2, self.ec, self.front_ec_i)

        #finding optimal outside turn path
        goingLeft = self.ec.is_next_Left(self.front_ec_i)
        if goingLeft:
            (to_point_optimal_outside, optimal_outside_theta1, optimal_outside_theta2) = calculate_steering(radians(MAX_RIGHT_ANGLE), radians(MAX_LEFT_ANGLE), dd, 10, OUTSIDE_TURN_ERROR, self.pos, self.theta1, self.theta2, self.ec, self.front_ec_i)
        else:
            (to_point_optimal_outside, optimal_outside_theta1, optimal_outside_theta2) = calculate_steering(radians(MAX_RIGHT_ANGLE), radians(MAX_LEFT_ANGLE), dd, 10, -OUTSIDE_TURN_ERROR, self.pos, self.theta1, self.theta2, self.ec, self.front_ec_i)


        #check if the nodes are within the allowed track and we haven't reached the maximum error
        #Left
        (inTrack, truck_error, b_ec_i, f_ec_i) = self.trackChecker.checkIfInTrack(self.pos, self.theta1, self.theta2, to_point_left, left_theta1, left_theta2, self.dt, self.ec, self.front_ec_i, self.back_ec_i)
        if inTrack and self.totError+truck_error <= self.lowest_error:
            self.addState(to_point_left, left_theta1, left_theta2, truck_error, b_ec_i, f_ec_i)
        #Strait
        (inTrack, truck_error, b_ec_i, f_ec_i) = self.trackChecker.checkIfInTrack(self.pos, self.theta1, self.theta2, to_point_strait, strait_theta1, strait_theta2, self.dt, self.ec, self.front_ec_i, self.back_ec_i)
        if inTrack and self.totError+truck_error <= self.lowest_error:
            self.addState(to_point_strait, strait_theta1, strait_theta2, truck_error, b_ec_i, f_ec_i)
        #Right
        (inTrack, truck_error, b_ec_i, f_ec_i) = self.trackChecker.checkIfInTrack(self.pos, self.theta1, self.theta2, to_point_right, right_theta1, right_theta2, self.dt, self.ec, self.front_ec_i, self.back_ec_i)
        if inTrack and self.totError+truck_error <= self.lowest_error:
            self.addState(to_point_right, right_theta1, right_theta2, truck_error, b_ec_i, f_ec_i)
        #Optimal outside turn
        (inTrack, truck_error, b_ec_i, f_ec_i) =self.trackChecker.checkIfInTrack(self.pos, self.theta1, self.theta2, to_point_optimal_outside, optimal_outside_theta1, optimal_outside_theta2, self.dt, self.ec, self.front_ec_i, self.back_ec_i)
        if inTrack:
            self.addState(to_point_optimal_outside, optimal_outside_theta1, optimal_outside_theta2, truck_error, b_ec_i, f_ec_i)
        #Optimal
        (inTrack, truck_error, b_ec_i, f_ec_i) = self.trackChecker.checkIfInTrack(self.pos, self.theta1, self.theta2, to_point_optimal, optimal_theta1, optimal_theta2, self.dt, self.ec, self.front_ec_i, self.back_ec_i)
        if inTrack and self.totError+truck_error <= self.lowest_error:
            self.addState(to_point_optimal, optimal_theta1, optimal_theta2, truck_error, b_ec_i, f_ec_i)


    def gatherPath(self, startPoint, endPoint, end_theta1, end_theta2, end_err, err):
        path = []
        fromPoints = {}
        self.fromPoints[(endPoint.x, endPoint.y)] = ((self.pos.x, self.pos.y),self.theta1, self.theta2, err)
        fromPoints[(endPoint.x, endPoint.y)] = ((self.pos.x, self.pos.y),self.theta1, self.theta2, err)
        prex = self.pos.x
        prey = self.pos.y
        pret1 = self.theta1
        pret2 = self.theta2
        prerr = err
        while not (prex== startPoint.x and  prey == startPoint.y):
            path.append(VehicleState(prex,prey, pret1, pret2))
            ((nx,ny),nt1, nt2, err) = self.fromPoints[prex,prey]
            fromPoints[(prex, prey)] = ((nx,ny),nt1, nt2, err)
            prex=nx
            prey=ny
            pret1 = nt1
            pret2 = nt2
            prerr = err
        return (fromPoints, list(reversed(path)), self.ec)

    def gather_x_y_path(self, startPoint, endPoint, end_theta1, end_theta2, err):
        path = []
        self.fromPoints[(endPoint.x, endPoint.y)] = ((self.pos.x, self.pos.y),self.theta1, self.theta2, err)
        prex = self.pos.x
        prey = self.pos.y
        while not (prex== startPoint.x and  prey == startPoint.y):
            path.append(Position(prex,prey))
            ((nx,ny),nt1, nt2, err) = self.fromPoints[prex,prey]
            prex=nx
            prey=ny
        return list(reversed(path))


    def addState(self, point, th1, th2, error, b_ec_i, f_ec_i):
        #add the vector as an adjacent vector to the previous vector in the graph
        self.fromPoints[(point.x, point.y)] = ((self.pos.x, self.pos.y),self.theta1, self.theta2, error)
        self.toVisit.append(((point.x,point.y), th1, th2, error, self.totError + abs(error), b_ec_i, f_ec_i))
        ((round_x, round_y), round_th1, round_th2) = rounding(point.x, point.y, th1, th2)
        self.errorList[((round_x, round_y), round_th1, round_th2)] = self.totError + abs(error)

File: scripts/recalculatePath.cpp
#include "recalculatePath.hpp"

RecalculatePath::RecalculatePath(TrackChecker *track_checker_){
    track_checker = track_checker_;
}

std::list<VehicleState*> RecalculatePath::calculate_path(VehicleState *startVs, Point *endPoint, Point *secondEndPoint
    , double totError_, ErrorCalc *front_ec_, double modPoint_, double modTheta_){

    std::cout << "startError: " << totError_ << std::endl;

    modPoint = modPoint_;
    modTheta = modTheta_;
    theta1 = startVs->th1;
    theta2 = startVs->th2;
    front_ec = front_ec_;
    back_ec = front_ec->getCopy();
    pos = new Point(startVs->x, startVs->y);
    lowest_error = totError_;
    totError = 0;

    //Reset path's
    //TODO: Remove before finishing
    toVisit = new std::stack<VehicleState_error*>();
    visited = new std::unordered_set<VehicleState>();
    fromPoints = new std::unordered_map<Point , VehicleState>();
    errorList = new std::unordered_map<VehicleState , double>();
    path = new std::list<VehicleState*>();

    //add all possible pathes for the first point before looping
    addPossiblePathes();

    while(toVisit->size()>0){
        VehicleState_error *new_visit;
        //loop until all possible nodes have been visited
        while(1){
            if(toVisit->size()== 0){
                //reached end
                std::cout << "reached end" << std::endl;
                return *path;
            }
            new_visit = toVisit->top();
            toVisit->pop();
            //msg.x = new_visit->vs->x;
            //msg.y = new_visit->vs->y;
            //visited_pub.publish(msg);
            //round to not having to visit every mm, to make it faster
            VehicleState *rounded_vs = rounding(new_visit->vs, modPoint, modTheta);
            //find the previous error for this grid
            auto iter_error_list = errorList->find(*rounded_vs);
            double prev_err = iter_error_list->second;

            //visit the node if we haven't visited it before and we have not reached the upper bound, or visit if we reached same grid with lower error
            auto iter_visited = visited->find(*rounded_vs);
            if((iter_visited == visited->end() && new_visit->totError < lowest_error) || prev_err > new_visit->totError){
                break;
            }
            else{
                delete new_visit;
            }
        }
        double err = new_visit->error;
        //found new node to visit
        pos->x = new_visit->vs->x;
        pos->y = new_visit->vs->y;
        theta1 = new_visit->vs->th1;
        theta2 = new_visit->vs->th2;
        front_ec = new_visit->front_ec;
        back_ec = new_visit->back_ec;
        totError = new_visit->totError;
        delete new_visit;

        //std::cout << "visiting x: " << pos->x << "y: " << pos->y << std::endl;

        double dist = sqrt((endPoint->x - pos->x)*(endPoint->x - pos->x) + (endPoint->y - pos->y)*(endPoint->y - pos->y));
        if(front_ec->isAboveEnd(secondEndPoint,endPoint, pos->x, pos->y) && dist <1*dt && front_ec->isAtEnd()){ //checks if we are above a line of the two last points
            //reached a sloution, gather the path if we are on the optimal path
            double dd = speed * dt;
            VehicleState *cur_vs = new VehicleState(pos->x, pos->y, theta1, theta2);
            VehicleState *check_vs = calculate_steering(radians(MAX_LEFT_ANGLE), radians(MAX_RIGHT_ANGLE), dd, 10, 0, cur_vs, front_ec);
            delete cur_vs;
            double nerror = front_ec->calculateError(check_vs->x, check_vs->y);
            if(abs(nerror)< 1 && totError < lowest_error){
                std::cout << "Found better solution, error: " << totError << std::endl;
                Point *new_point = new Point(check_vs->x, check_vs->y);
                Point *startPoint = new Point(startVs->x, startVs->y);
                path = gatherPath(startPoint, new_point);
                lowest_error = totError;
            }
            //Mark visited
            VehicleState *in_r_vs = new VehicleState(pos->x, pos->y, theta1, theta2);
            VehicleState *rounded_vs = rounding(in_r_vs, modPoint, modTheta);
            delete in_r_vs;
            visited->insert(*rounded_vs);

        } else {
            //we have not yet visited all nodes, search for new possible nodes
            addPossiblePathes();

            //round to not having to visit every mm, for making it faster
            VehicleState *in_r_vs = new VehicleState(pos->x, pos->y, theta1, theta2);
            VehicleState *rounded_vs = rounding(in_r_vs, modPoint, modTheta);
            delete in_r_vs;
            //mark the previous node/state as visited
            visited->insert(*rounded_vs);
        }
    }
    //reached end
    std::cout << "reached end" << std::endl;
    return *path;
}

void RecalculatePath::addPossiblePathes(){
    double dd = speed * dt;
    //add all possible pathes from pos, theta1 and theta2
    double steering_angle_rad = 0;
    VehicleState *current_vs = new VehicleState(pos->x, pos->y, theta1, theta2);

    VehicleState *strait_vs = calculateNextState(current_vs, dd, steering_angle_rad);
    //going right
    steering_angle_rad = radians(MAX_RIGHT_ANGLE); //max right angle
    VehicleState *right_vs = calculateNextState(current_vs, dd, steering_angle_rad);
    //going left
    steering_angle_rad = radians(MAX_LEFT_ANGLE); //max left angle
    VehicleState *left_vs = calculateNextState(current_vs, dd, steering_angle_rad);
    //finding optimal path
    VehicleState *optimal_vs = calculate_steering(radians(MAX_LEFT_ANGLE), radians(MAX_RIGHT_ANGLE), dd, 10, 0, current_vs, front_ec);
    //Optimal outside turn
    bool goingLeft = front_ec->is_next_left();
    VehicleState *optimal_outside_vs;
    if (goingLeft){
        optimal_outside_vs = calculate_steering(radians(MAX_LEFT_ANGLE), radians(MAX_RIGHT_ANGLE), dd, 10, OUTSIDE_TURN_ERROR, current_vs, front_ec);
    } else {
        optimal_outside_vs = calculate_steering(radians(MAX_LEFT_ANGLE), radians(MAX_RIGHT_ANGLE), dd, 10, -OUTSIDE_TURN_ERROR, current_vs, front_ec);
    }
    delete current_vs;

    //Left
    Point *left_point = new Point(left_vs->x, left_vs->y);
    InTrack * left_it = track_checker->checkIfInTrack(pos, theta1, theta2, left_point, left_vs->th1, left_vs->th2, front_ec, back_ec);
    if(left_it->in_track && totError + left_it->error <= lowest_error){
        addState(left_point, left_vs->th1, left_vs->th2, left_it->error);
    } else {delete left_point;}
    delete left_vs;

    //Strait
    Point *strait_point = new Point(strait_vs->x, strait_vs->y);
    InTrack * strait_it = track_checker->checkIfInTrack(pos, theta1, theta2, strait_point, strait_vs->th1, strait_vs->th2, front_ec, back_ec);
    if (strait_it->in_track && totError + strait_it->error <= lowest_error){
        addState(strait_point, strait_vs->th1, strait_vs->th2, strait_it->error);
    }else { delete strait_point;}
    delete strait_vs;

    //Right
    Point *right_point = new Point(right_vs->x, right_vs->y);
    InTrack *right_it = track_checker->checkIfInTrack(pos, theta1, theta2, right_point, right_vs->th1, right_vs->th2, front_ec, back_ec);
    if(right_it->in_track && totError + right_it->error <= lowest_error){
        addState(right_point, right_vs->th1, right_vs->th2, right_it->error);
    }else {delete right_point;}
    delete right_vs;

    //Optimal outside turn
    Point *optimal_outside_point = new Point(optimal_outside_vs->x, optimal_outside_vs->y);
    InTrack * optimal_outside_it = track_checker->checkIfInTrack(pos, theta1, theta2, optimal_outside_point, optimal_outside_vs->th1, optimal_outside_vs->th2, front_ec, back_ec);
    if (optimal_outside_it->in_track && totError + optimal_outside_it->error <= lowest_error){
        addState(optimal_outside_point, optimal_outside_vs->th1, optimal_outside_vs->th2, optimal_outside_it->error);
    } else {delete optimal_outside_point;}
    delete optimal_outside_vs;

    //Optimal
    Point *optimal_point = new Point(optimal_vs->x, optimal_vs->y);
    InTrack * optimal_it = track_checker->checkIfInTrack(pos, theta1, theta2, optimal_point, optimal_vs->th1, optimal_vs->th2, front_ec, back_ec);
    if(optimal_it->in_track && totError + optimal_it->error <= lowest_error){
        addState(optimal_point, optimal_vs->th1, optimal_vs->th2, optimal_it->error);
    } else {delete optimal_point;}
    delete optimal_vs;

}

std::list<VehicleState*>* RecalculatePath::gatherPath(Point *startPoint, Point *endPoint){
    std::list<VehicleState*>* path = new std::list<VehicleState*>();
    //fromPoints.insert({*endPoint, new VehicleState(pos->x, pos->y, theta1, theta2)});
    double prex = pos->x;
    double prey = pos->y;
    double pret1 = theta1;
    double pret2 = theta2;
    Point *f_point;
    while (!(prex== startPoint->x && prey == startPoint->y)){
        path->push_front(new VehicleState(prex, prey, pret1, pret2));
        f_point = new Point(prex,prey);
        auto iter = fromPoints->find(*f_point);
        VehicleState vs = iter->second;
        prex=vs.x;
        prey=vs.y;
        pret1 = vs.th1;
        pret2 = vs.th2;
        delete f_point;
    }
    return path;
}

void RecalculatePath::addState(Point *point, double th1, double th2, double error){
    //add the point as an adjacent point
    //TODO: shuld not be same error for fromPoint and toVisit?
    VehicleState *vs = new VehicleState(pos->x, pos->y, theta1, theta2);
    fromPoints->insert({*point, *vs});
    VehicleState_error *vs_to = new VehicleState_error(new VehicleState(point->x, point->y, th1, th2), error, front_ec->getCopy(), back_ec->getCopy());
    double total_error = totError + abs(error);
    vs_to->totError = total_error;
    toVisit->push(vs_to);

    VehicleState *in_r_vs = new VehicleState(point->x, point->y, th1, th2);
    VehicleState *rounded_vs = rounding(in_r_vs, modPoint, modTheta);
    errorList->insert({*rounded_vs, total_error});
    delete in_r_vs;

}
